### PR TITLE
[WIP] Add support for random access to backend objects, not fetching entire objects.

### PR DESCRIPTION
**Pull request by pontus**
**Related issue(s) and PR(s)**  
This helps parts for #363 and #696.

### CODE REVIEW

**Comment by pontus** [comment]
I've done some performance tests with this code, just downloading a dataset shared with someone.

With the old s3 driver (leaving it mostly up to aws sdk), it downloads 87239339037 bytes in 4 minutes 36s (~ 316 Mbyte/second). 

After adding some performance optimisations I see the same takes 6 minutes 19 seconds (~219 Mbyte/second) with the seekable driver (a lot of that comes from needing to fail if the object doesn't exist/can't be accessed, which makes me do a `HEAD` call).

File: sda-download/internal/storage/storage.go [hunk 1]
```diff
@@ -140,6 +167,15 @@ type S3Conf struct {
...
 
+func newS3SeekableBackend(config S3Conf) (*s3SeekableBackend, error) {
+	sb, err := newS3Backend(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &s3SeekableBackend{*sb}, nil
+}
+
```
Comment (jbygdell) [inline sda-download/internal/storage/storage.go:178]: Is a dedicated backed really needed?
It is after all just a duplication of the existing backend.

File: sda-download/internal/storage/storage.go [hunk 2]
```diff
@@ -315,3 +351,439 @@ func transportConfigS3(config S3Conf) http.RoundTripper {
...
+		filePath,
+		objectSize,
+		sync.Mutex{},
+		make([]int64, 0, 32),
+		false,
+		nil,
+	}
+
+	return reader, nil
+}
```
Comment (jbygdell) [inline sda-download/internal/storage/storage.go:376]: Wouldn't the implementation of this as a `SeekableFileReader` in the normal backend be the same, so we can do both types of file reads depending on the incoming query.

File: sda-download/internal/storage/storage.go (hunk 1 above)
Comment (pontus) [reply sda-download/internal/storage/storage.go:178]: Correct.

I made it as a separate to be able to possibly merge it but still use the previous functionality only if you do not feel confident with it (e.g. your use cases don't care about delivering parts and you prefer higher performance). Can also be imlemented with a parameter to control behaviour of the driver.

File: sda-download/internal/storage/storage.go (hunk 2 above)
Comment (pontus) [reply sda-download/internal/storage/storage.go:376]: True. I made this in a way to keep the `s3` driver unchanged for now, but in an ideal future scenario, I would like to see a single driver (supporting seeking), possibly offering some more controls to adjust how it works.

**Comment by MalinAhlberg** [comment]
I tested this with htsget, using [this branch](https://github.com/neicnordic/sensitive-data-archive/commit/e947be245e8a2e33b6617c92d38254d7cc566603) (wip, not cleaned-up) . Everything seems to work well!

**Comment by pontus** [comment]
See #826

**Pull request closed without merging**